### ISSUE TITLE
Issue with monai handler MeanDice while doing multi-label sementation

### ISSUE THREAD

**Issue by saruarlive** [Project-MONAI/MONAI#1306]
Dear @Nic-Ma et al,

Hope you are doing well. 

Binary segmentation works as expected while using monai/ignite module. 
While doing multi-label segmentation using monai/ignite module, it seems that MeanDice handler is not giving expected mean dice score as it is shown below,

_INFO:ignite.engine.engine.UNetEvaluator:Epoch[20] Metrics -- val_mean_dice: 0.0108 
INFO:ignite.engine.engine.UNetEvaluator:Key metric: val_mean_dice best value: 0.029973947008450828 at epoch: 8

INFO:ignite.engine.engine.UNetEvaluator:Epoch[90] Metrics -- val_mean_dice: 0.0005 
INFO:ignite.engine.engine.UNetEvaluator:Key metric: val_mean_dice best value: 0.029973947008450828 at epoch: 8_

When I tried with native PyTorch training loop using the same data loader, I use to get different mean dice score, starting from 0.1-0.75. 

What is the issue here? Could you kindly give any solution so that I can use monai/ignite engine module for multi-label segmenation? I am looking forward to hearing from you.

**Comment by Nic-Ma** [comment]
Hi @saruarlive ,

You can refer to this example: https://github.com/Project-MONAI/tutorials/blob/master/modules/engines/unet_training_dict.py
Even though it's NOT a multi-labels example.
I think probably your issue came from the `loss function` or `post_transforms`, please make sure you set expected `activation layer` for loss function and set `activation layer` & `AsDiscreted` transform for the model predictions before metrics.

Thanks.

**Comment by saruarlive** [comment]
Dear @Nic-Ma et al,
Thanks for your reply.
I saw that like earlier. Still, I could not figure what is the issue, here is my loss function/post transforms

_loss function = DiceLoss(to_onehot_y=False, sigmoid=True, squared_pred=True)

val_post_transform = Compose(
    [
        Activationsd(keys = ('pred', 'label'), sigmoid=(True, False)), 
        AsDiscreted(keys=("pred", "label"), threshold_values=(True, False)),  ###logit_thresh=(0.5, 0.0)
    ]
)_

**Comment by saruarlive** [comment]
Dear  @Nic-Ma  et al,

The problem with the MeanDice handler still exists for multi-label segmentation. The training and validation accuracy is working. Any suggestion on how to fix the issue. 
 
![image](https://user-images.githubusercontent.com/24542592/101174543-506e1600-3644-11eb-8090-c42a6ae4d5e2.png)

**Comment by Nic-Ma** [comment]
Hi @saruarlive ,

Your code looks good to me, could you please also share your PyTorch program to compare it?
And it's better to paste your whole MONAI workflow program for analysis.
@ericspod @wyli could you please help also take a look where it is wrong?

Thanks.

**Comment by saruarlive** [comment]
Dear @Nic-Ma @wyli @ericspod, I am sharing ignite code since my problem with ignite handler only,

**In the evaluation section:**

```javascript

val_handlers = [
    ProgressBar(),
    StatsHandler(output_transform=lambda x: None),
    CheckpointSaver(save_dir="./segResult/", save_dict={"net": model}, save_key_metric=True, key_metric_n_saved=3),
    #CheckpointSaver(save_dir=model_folder, save_dict={"net": net}, save_key_metric=True, key_metric_n_saved=3),
]


val_post_transform = Compose(
    [
        Activationsd(keys="pred", sigmoid=True),
        AsDiscreted(keys="pred", threshold_values=True),
        #KeepLargestConnectedComponentd(keys="pred", applied_labels=[1]),
    ]
)

# Define customized evaluator
class UNetEvaluator(SupervisedEvaluator):
    def _iteration(self, engine, batchdata):
        inputs, targets = self.prepare_batch(batchdata)
        inputs, targets = inputs.to(engine.state.device), targets.to(engine.state.device)
        flip_inputs = torch.flip(inputs, dims=(2, 3, 4))

        def _compute_pred():
            pred = self.inferer(inputs, self.network)
            flip_pred = torch.flip(self.inferer(flip_inputs, self.network), dims=(2, 3, 4))
            return (pred + flip_pred) / 2
        
        # execute forward computation
        self.network.eval()
        with torch.no_grad():
            if self.amp:
                with torch.cuda.amp.autocast():
                    predictions = _compute_pred()
            else:
                predictions = _compute_pred()
        return {"image": inputs, "label": targets, "pred": predictions}

#dice_metric = DiceMetric(include_background=True, reduction="mean")
evaluator = UNetEvaluator(
    device=device,
    val_data_loader=val_loader,
    network=model,
    inferer=SimpleInferer(),
    post_transform=val_post_transform,
    key_val_metric={"val_mean_dice": MeanDice(include_background=True, output_transform=lambda x: (x["pred"], x["label"]))},
    additional_metrics={"val_acc": Accuracy(output_transform=lambda x: (x["pred"], x["label"]))},
    val_handlers=val_handlers,
    amp=True,
)

```

**in the training section**

```javascript

train_handlers = [
    #ProgressBar(),
    LrScheduleHandler(lr_scheduler=scheduler, print_lr=True),
    ValidationHandler(validator=evaluator, interval=2, epoch_level=True),
    StatsHandler(tag_name="train_loss", output_transform=lambda x: x["loss"]),
]

train_post_transforms = Compose(
    [
        Activationsd(keys="pred", sigmoid=True),
        AsDiscreted(keys="pred", threshold_values=True),
        #KeepLargestConnectedComponentd(keys="pred", applied_labels=[1]),
    ]
)


# define customized trainer

class UNetTrainer(SupervisedTrainer):
        
    def _iteration(self, engine, batchdata):
        
        inputs, targets = self.prepare_batch(batchdata)
        inputs, targets = inputs.to(engine.state.device), targets.to(engine.state.device)

        
        def _compute_loss(preds, label): 
            return self.loss_function(preds, label)
    

        self.network.train()
        self.optimizer.zero_grad()
        if self.amp and self.scaler is not None:
            with torch.cuda.amp.autocast():
                predictions = self.inferer(inputs, self.network)
                loss = _compute_loss(predictions, targets)
            self.scaler.scale(loss).backward()
            self.scaler.step(self.optimizer)
            self.scaler.update()
        else:
            predictions = self.inferer(inputs, self.network)
            loss = _compute_loss(predictions, targets).mean()
            loss.backward()
            self.optimizer.step()
        return {"image": inputs, "label": targets, "pred": predictions, "loss": loss.item()}


trainer = UNetTrainer(
    device=device,
    max_epochs=max_epochs,
    train_data_loader=train_loader,
    network=model,
    optimizer=optimizer,
    loss_function=loss_function,
    inferer=SimpleInferer(),
    post_transform=train_post_transforms,
    key_train_metric={"train_acc": Accuracy(output_transform=lambda x: (x["pred"], x["label"]))},
    train_handlers=train_handlers,
    amp=True,
)

```
### Finally running,
```javascript
import logging
import sys
logging.basicConfig(stream=sys.stdout, level=logging.INFO)

trainer.run()
```
After 100 hundred epochs, I got logs as shown below,

![image](https://user-images.githubusercontent.com/24542592/101228432-fdc04880-369b-11eb-9cb4-ee5731b17bed.png)

What I am missing here? I suppose to get a mean dice score around 0.80. I am looking forward to hearing from you.

**Comment by wyli** [comment]
just wanted to make it clear that we won't be able to provide you deep learning consultancy just for your personal projects. but if you could provide sufficient information that is potentially beneficial for the other users and the community, we'll be happy to help.
